clusterapi scale from zero support

elmiko opened the pull request:
#### Which component this PR applies to? 

cluster-autoscaler

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR adds scale from zero capability to the clusterapi provider for cluster-autoscaler.

#### Which issue(s) this PR fixes:

Fixes #3150 

#### Special notes for your reviewer:

The [cluster-api-provider-kubemark](https://github.com/kubernetes-sigs/cluster-api-provider-kubemark) has implemented the provider part of the opt-in API for scaling from zero. I recommend using this as a way to test with a live cluster.

#### Does this PR introduce a user-facing change?

```release-note
Added support for scaling to and from zero nodes for the cluster autoscaler's Cluster API provider. Enabling this feature will require changes by the user, for instruction please see the Cluster API (clusterapi) provider README file in the autoscaler repository.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- [CAEP]  https://github.com/kubernetes-sigs/cluster-api/blob/release-1.1/docs/proposals/20210310-opt-in-autoscaling-from-zero.md

Usage docs are included in the README file.

elmiko commented:
i am placing a pre-empitve hold on this to allow time for reviews.
/hold for reviews

@enxebre ptal

jackfrancis commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go:
```diff
@@ -226,7 +234,64 @@ func (ng *nodegroup) Nodes() ([]cloudprovider.Instance, error) {
 // Implementation optional.
 func (ng *nodegroup) TemplateNodeInfo() (*schedulerframework.NodeInfo, error) {
-	return nil, cloudprovider.ErrNotImplemented
+	if !ng.scalableResource.CanScaleFromZero() {
```
💥

jackfrancis commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go:
```diff
@@ -226,7 +234,64 @@ func (ng *nodegroup) Nodes() ([]cloudprovider.Instance, error) {
+	_, cpuPresent := capacity[corev1.ResourceCPU]
+	_, memPresent := capacity[corev1.ResourceMemory]
+	if !cpuPresent || !memPresent {
+		return nil, cloudprovider.ErrNotImplemented
```
is `cloudprovider.ErrNotImplemented` the right error type to return for this error condition?

jackfrancis replied:
After reading through [this](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210310-opt-in-autoscaling-from-zero.md#implementation-detailsnotesconstraints) it seems like the answer is yes.

jackfrancis replied:
Actually now I see that the `CanScaleFromZero()` method already validates the presence of cpu and memory capacity:

https://github.com/kubernetes/autoscaler/pull/4840/files#diff-2052c87a1cd8247c67b54b3e35acc9dff157210e845256b2282f87ac46bf805fR205

So I think we can remove L246-250 entirely

elmiko replied:
removed the extra lines and added some code docs

jackfrancis commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go:
```diff
@@ -226,7 +234,64 @@ func (ng *nodegroup) Nodes() ([]cloudprovider.Instance, error) {
+			labels = cloudprovider.JoinStringMaps(labels, extractNodeLabels(node))
+		}
+	}
+	return labels, nil
```
reading through the various labels funcs here it seems like there will never be a situation where the labels `map[string]string` will be nil (at present `buildGenericLabels` will always guarantee a bunch of labels), but do we want to be a bit more defensive about that and return an error if for any reason our `labels` var ends up nil at the end of this function?

elmiko replied:
i need to think that through a little bit more, but it sounds like a good defensive practice. the idea would be that if we don't see the labels we expect, then we would return some sort of error (which would cause the scaling to not happen) ?

jackfrancis replied:
It's certainly valid for the `nodeInfo` object (type `*schedulerframework.NodeInfo
`) to not have labels.

In any event this is probably academic as the UT in `cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go` will protect us from this scenario. I think we're fine.

elmiko replied:
> It's certainly valid for the nodeInfo object (type *schedulerframework.NodeInfo ) to not have labels.

agreed, my concern is messing up the simulator logic that uses the labels for matching. i'll do a little more investigation.

elmiko commented:
updated to remove redundant checks for cpu and memory data

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go:
```diff
@@ -1181,16 +1239,252 @@ func TestNodeGroupWithFailedMachine(t *testing.T) {
+				},
+				nodegroupLabels: map[string]string{
+					"nodeGroupLabel": "value",
+					"anotherLabel":   "anotherValue",
```
Might be more obvious if this was as below, took me a moment to work out why it was another value and not the node value
```suggestion
					"anotherLabel":   "nodeGroupValue",
```

elmiko commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -53,14 +55,29 @@ const (
+// infrastructureTemplateCache holds infrastructure machine templates and
+// their time to live values. It is used by the getInfrastructureResource
+// function.
+var infrastructureTemplateCache map[string]infrastructureTemplateCacheEntry = map[string]infrastructureTemplateCacheEntry{}
```
imo, we want it global because these records will be shared by all possible threads, although i have a feeling we won't hit an issue.

JoelSpeed commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -708,3 +725,47 @@ func (c *machineController) allowedByAutoDiscoverySpecs(r *unstructured.Unstruct
+	}
+
+	// update the cache with the new record
+	infrastructureTemplateCache[cachekey] = infrastructureTemplateCacheEntry{
```
Perhaps a sync.Map might be the easiest things to use?

enxebre commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -708,3 +725,47 @@ func (c *machineController) allowedByAutoDiscoverySpecs(r *unstructured.Unstruct
+			return entry.template, nil
+		}
+	}
+
```
>because of the frequency of read requests

It's not clear to me why we need this handcrafted cache, could you elaborate on the comment above?

Shouldn't the managementClient just be told what it caches and what it doesn't via e.g controller runtime primitives? e.g
```
NewCache: cache.BuilderWithOptions(cache.Options{
			SelectorsByObject: cache.SelectorsByObject{
...
```

elmiko replied:
good question. i initially tried adding the infrastructure template types to our dynamic client, but what i noticed when running that way is that we get client side throttled several times per sync loop. given that these records will not update very frequently, if at all, it seemed like a better performance to use a local cache rather than attempting to hit the api server so many times.

the reason for why we are hitting that api server so much is that we check for the scale from zero capability and limits many times within a single sync interval. if the community here is ok with our logs being full of client side throttling messages i can certainly switch this back to the other style, but the cache seemed to give a better user experience from the log perspective.

enxebre replied:
>when running that way is that we get client side throttled several times per sync loop

Thanks for context Mike! Can you help me understand why de above happens? How's this different than any other resource we are consuming? why is a hand crafted cache justified only here? I think to mitigate overloading the api server we should be able to use existing kube core client cache mechanisms aka informers/indexers and if needed controller runtime abstractions . I'd be reluctant to have our own cache unless there's a strong reason for it.

elmiko replied:
yeah, it's a great point Alberto, and i am not overly fond of creating a cache here but it seemed like the best way to handle this for the infrastructure templates only.

my rationale goes along these lines,
1. we need to discover the infrastructure template gvk and names from the machinesets and machinedeployments
2. adding the infrastructure templates to our dynamic informer caches does work, but given the frequency with which we check the scale from zero information, this produces a higher number of api calls than the calls to get the machinesets and machinedeployments
3. we expect a `capacity` field in the `status` for the infrastructure templates, in practice this capacity should not change for the life of the template. that said, it is possible that there might be a a period of time between when an infrastructure template is created and when the actuator adds the capacity information, so we need to be able to reconcile these just in case it does change.
4. given the statements above, it seems more reasonable to cache the information from the template rather than request new information on every sync interval, especially when multiple machinesets/machinedeployments could share the same template.

that's my thinking, if we all agree that the excess calls to the api server are ok, then i will change it back to my original design. but, i do feel that the local cache is appropriate here.

enxebre replied:
>given the statements above, it seems more reasonable to cache the information from the template rather than request new information on every sync interval, especially when multiple machinesets/machinedeployments could share the same template.
that's my thinking, if we all agree that the excess calls to the api server are ok, then i will change it back to my original design. but, i do feel that the local cache is appropriate here.

I agree with the above, what I'm not seeing is why you need a handcrafted cache for that and to avoid excess calls to the api server rather than use built-in caches as everywhere else. Kube core go client and controller runtime give you caches that we already use here and widely in the ecosystem. Why can you use kubernetes informer/indexer and their higher abstractions with the resync config of your choice as usual?

elmiko replied:
gonna reverse the order of your questions ;)

> Why can't you use kubernetes informer/indexer and their higher abstractions with the resync config of your choice as usual?

as far as i know nothing aside from my lack of knowledge around how to craft those caches. it wasn't clear to me how i would construct a cache through go-client where i could reduce the sync period. when i add the template types to the current client as we discover them, i see a lot of noise in the logs from the client blocking the calls due to the rate at which we are making calls. if those logs messages are acceptable, i will put the original client-go cache back in place.

> what I'm not seeing is why you need a handcrafted cache for that and to avoid excess calls to the api server rather than use built-in caches as everywhere else.

it was to reduce the log noise. using the built-in caches produces a lot of noise in the logs about `Waited for XXXms due to client-side throttling....`. is there a way to control how frequently the client-go client will connect with the API server?


and just to be clear, this is not a blocker for me. if we agree that the log noise is acceptable, i will revert to use the client-go cache.

elmiko replied:
just to try and add a little more explanation,

the reason i felt the noise logs were an issue we should deal with is that we've had users question how the capi provider works with respect to seeing too many throttling messages in the logs. i don't want our users to think there is some problem that we could alleviate by reducing the frequency of our checks.

enxebre replied:
If you are seeing `XXXms due to client-side throttling.....` every loop I question it's actually using the cache. This i happening now for e.g get scale subresource. See https://github.com/kubernetes/autoscaler/pull/5025

To move forward with this PR please let's move towards using native client/informers for cache.

elmiko replied:
> If you are seeing `XXXms due to client-side throttling.....` every loop I question it's actually using the cache. This i happening now for e.g get scale subresource. See #5025

when i add my cache most of those messages go away for the infra templates

> To move forward with this PR please let's move towards using native client/informers for cache.

i'm ok with removing the additional cache, it just means our logs get more noisy. if that becomes an issue we can always put it back in place.

thanks for the followup =)

enxebre replied:
>when i add my cache most of those messages go away for the infra templates

If you see it without your cache is because that client is hitting the api server instead of caching and that can be just codify to use the cache instead either in this PR or we can follow up later, let's just make sure we capture it in an issue.

elmiko commented:
update
* rebased
* removed infrastructure machine template cache
* improve unit tests based on review comments

i am doing some manual testing now to make sure nothing broke, but i think this is ready for re-review

elmiko commented:
manual testing is working as expected against capi 1.0.5 and kubemark 0.3.0

enxebre commented:
> manual testing is working as expected against capi 1.0.5 and kubemark 0.3.0
> I've given this a review again and it LGTM, do we want to give others a few days to review and then try to get this merged? Is there anyone specific we want to get to review this before we label it?

sgtm, though we definitely need to add some sort of smoke testing to validate PRs here.

enxebre commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -708,3 +720,21 @@ func (c *machineController) allowedByAutoDiscoverySpecs(r *unstructured.Unstruct
+
+// Get an infrastructure machine template given its GVR, name, and namespace.
+func (c *machineController) getInfrastructureResource(resource schema.GroupVersionResource, name string, namespace string) (*unstructured.Unstructured, error) {
+	infra, err := c.managementClient.
```
The logs you mentioned are because of this client is not using a cache. We should be using a native cache e.g https://github.com/kubernetes-sigs/controller-runtime/blob/88234a8fc9ad7f559a593e098f224dcecf4c6f6f/pkg/cache/internal/informers_map.go#L311

controller runtime gives you the abstractions for it https://github.com/kubernetes-sigs/controller-runtime/issues/615

elmiko replied:
ok, perhaps i'm just not understanding the client well enough. i had thought that by using the dynamic client and adding these types as we discover them that they will be added to an internal cache.

also, we are using client-go afaik and not controller-runtime in here. i would be happy to fix this up so that we are not overloading the kas, i just need some pointers about how to get there using client-go.

elmiko replied:
ok, think i've got this fixed

> It's not about having noisy logs or not,

fwiw, i agree, logs were just a symptom

enxebre commented:
thanks a lot @elmiko, this looks great to me other than https://github.com/kubernetes/autoscaler/pull/4840#discussion_r925771269

elmiko commented on cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:
```diff
@@ -53,14 +55,24 @@ const (
+// infrastructureTemplateCacheEntry holds a single infrastructure machine
+// template and a time to live. It is used by the infrastructure template cache
+// to hold entries.
+type infrastructureTemplateCacheEntry struct {
```
oh yeah, my bad.

elmiko commented:
fyi, i'm leaving this on hold so we don't merge early since it's already approved. i will remove the hold once we work out the last discussion item.

elmiko commented:
update
* removed all references to the custom cache
* squashes down to 2 commits

elmiko commented:
i'm doing some digging around client-go, have found https://github.com/kubernetes/client-go/tree/master/tools/cache but i think we need to add a little more code to make it work. i was under the false impression that the cache was automatic with client-go. hopefully i'll have another patch soon with the cache added.

elmiko commented:
i talked with @JoelSpeed a little about client-go internals, i think what i need to add here is the ability to add new informers to the client when we observe new infrastructure machine templates. i'm doing a little more digging to see how we can do that with our current setup.

elmiko commented:
ok, i have finally understood the magical incantations for client-go to get the dynamic informers created and running for the infrastructure templates. i think this is good to merge whenever folks are happy with the code quality.

/hold cancel

JoelSpeed commented:
This looks good and I think we've addressed everyones comments. I know Mike has been testing this and was seeing it working. I think it's good to merge now.
/lgtm
